[Tracking] Circuit Optimizations and Fixes

d0cd opened the pull request:
This PR tracks the fixes and optimizations made to circuits during the calibration period.
- #2047 
- #2048 
- #2050 
- #2052 
- #2055 
- #2072 
- #2049 
- #2093 
- #2100

d0cd committed:
Clean up mul_checked

d0cd committed:
Optimize mul_checked

d0cd committed:
Clean up mul_wrapped

d0cd committed:
Optimize pow_checked

d0cd committed:
Add TODO

d0cd committed:
Optimize unsigned checked addition

d0cd committed:
WIP fix for non-det bug in sqrt circuit

d0cd committed:
Fix uniqueness check

d0cd committed:
Fix console implementation, update constraint counts

d0cd committed:
Optimize shl_wrapped

d0cd committed:
Optimize unsigned shl_checked

d0cd committed:
Optimize signed shl_checked

d0cd committed:
Optimize shr_wrapped

acoglio committed:
Optimize field (in)equality circuit.

To calculate `z = x != y`, instead of
```
(z) (1 - z) = (0)
(x - y) (w) = (z)
(x - y) (1 - z) = (0)
```
(where `w` is an internal variable) we just need
```
(x - y) (w) = (z)
(x - y) (1 - z) = (0)
```
i.e. just the 2nd and 3rd constraints, because they imply the 1st one. So we
save one constraint for every field (in)equality.

This optimization has been already verified in the ACL2 theorem prover.

To avoid generating the boolean constraint, we need to avoid `Boolean::new`, and
instead generate the `Boolean` directly, as in boolean AND for example.
However, boolean AND has access to the private `LinearCombination` component of
`Boolean`, while field (in)equality does not. So I've added a `from_variable`
constructor to `Boolean`, but if there is a better or more desirable approach,
we can revise this part.

acoglio committed:
Shorten the field (in)equality argument.

acoglio committed:
Fix circuit to create group from coordinates.

This issue was discovered in the circuit to cast a field to a group, but it
really affects the sub-circuit of that circuit that turns two field coordinates
into a group element. The issue is not in the witness computation, but in the
circuit, which generates constraints about the point being in the group (i.e.
being on the curve, and being in the subgroup) that apply not to the given
point, but to a freshly created point, i.e. two different R1CS variables than
the ones given as input. This can be seen by generating sample circuits: there
is no connection between the input variables and the point being constrained.

This commit fixes the issue by adding equality constraints that tie the input
coordinates with the point that is constrained to be in the (sub)group. This is
not the most efficient circuit, because it uses more variables and constraints
than needed, but it is a simple fix for now.

d0cd committed:
Fix comment

howardwu committed:
Update circuit/types/integers/src/pow_checked.rs

Co-authored-by: Eric McCarthy <7607035+bendyarm@users.noreply.github.com>
Signed-off-by: Howard Wu <9260812+howardwu@users.noreply.github.com>

howardwu committed:
Update circuit/types/integers/src/pow_checked.rs

Co-authored-by: Eric McCarthy <7607035+bendyarm@users.noreply.github.com>
Signed-off-by: Howard Wu <9260812+howardwu@users.noreply.github.com>

d0cd committed:
Address feedback

d0cd committed:
Fix and update counts

d0cd committed:
Update circuit/types/integers/src/shl_checked.rs

Co-authored-by: Alessandro Coglio <coglio@kestrel.edu>
Signed-off-by: d0cd <pranavsaig@gmail.com>

d0cd committed:
Update circuit/types/integers/src/shl_checked.rs

Co-authored-by: Alessandro Coglio <coglio@kestrel.edu>
Signed-off-by: d0cd <pranavsaig@gmail.com>

d0cd committed:
Cleanup comment

d0cd committed:
Update circuit/types/integers/src/shr_wrapped.rs

Co-authored-by: Alessandro Coglio <coglio@kestrel.edu>
Signed-off-by: d0cd <pranavsaig@gmail.com>

d0cd committed:
Update circuit/types/integers/src/shr_wrapped.rs

Co-authored-by: Alessandro Coglio <coglio@kestrel.edu>
Signed-off-by: d0cd <pranavsaig@gmail.com>

d0cd committed:
Add comment

d0cd committed:
Use even square root in elligator

d0cd committed:
Update constraint counts

d0cd committed:
Refactor and fix console::elligator2::decode

d0cd committed:
Update counts

acoglio committed:
Use consistent comment style.

d0cd committed:
Update count

howardwu committed:
Merge pull request #2047 from AleoHQ/optimization/mul-checked

[Optimization] Checked Multiplication

howardwu committed:
Merge pull request #2048 from AleoHQ/optimize/add-checked

[Optimize] Unsigned checked addition

howardwu committed:
Merge pull request #2050 from AleoHQ/optimize/shl

[Optimize] Left shift.

howardwu committed:
Merge pull request #2052 from AleoHQ/optimize/shr

[Optimize] Right shift.

howardwu committed:
Merge branch 'tracking/circuit-optimizations-and-fixes' into feat/optimize-eq-circuits

howardwu committed:
Merge pull request #2055 from AleoHQ/feat/optimize-eq-circuits

Optimize field (in)equality circuit.

howardwu committed:
Update circuit/types/group/src/helpers/from_xy_coordinates.rs

Signed-off-by: Howard Wu <9260812+howardwu@users.noreply.github.com>

howardwu committed:
Merge pull request #2072 from AleoHQ/fix/cast-field-to-group

Fix circuit to create group from coordinates.

howardwu committed:
Merge branch 'tracking/circuit-optimizations-and-fixes' into fix/non-det-in-sqrt

d0cd committed:
Update counts

d0cd committed:
More updates

howardwu committed:
Fix doc comments in circuit impls

howardwu committed:
Merge pull request #2088 from AleoHQ/fix/doc-comments

Fix doc comments in circuit impls

howardwu committed:
Resolve merge

howardwu commented:
^@d0cd for the integer counts, I merged HEAD, however, `staging` might have some changes on `count()` logic. You can always force push back 1 commit if needed.

This is what I resolved manually:
```
Auto-merging synthesizer/process/src/tests/test_credits.rs
CONFLICT (content): Merge conflict in synthesizer/process/src/tests/test_credits.rs
Removing ledger/store/src/helpers/rocksdb/internal/iterator.rs
Auto-merging circuit/types/integers/src/shl_wrapped.rs
CONFLICT (content): Merge conflict in circuit/types/integers/src/shl_wrapped.rs
Auto-merging circuit/types/integers/src/shl_checked.rs
Auto-merging circuit/types/integers/src/pow_checked.rs
Auto-merging circuit/types/integers/src/mul_wrapped.rs
Auto-merging circuit/types/integers/src/mul_checked.rs
CONFLICT (content): Merge conflict in circuit/types/integers/src/mul_checked.rs
Auto-merging circuit/types/integers/src/div_checked.rs
CONFLICT (content): Merge conflict in circuit/types/integers/src/div_checked.rs
Auto-merging circuit/program/src/request/verify.rs
CONFLICT (content): Merge conflict in circuit/program/src/request/verify.rs
Automatic merge failed; fix conflicts and then commit the result.
```

acoglio committed:
Optimize some group circuits.

Refactor `Group::new` and `Group::from_xy_coordinates` to save constraints and
increase clarity. The enforcement of a point being in the (sub)group is now
delegated to a new circuit that does just that, and uses another new circuit to
enforce that a point is twice another (this differs from the circuit that
doubles a point).

This code has more separation between witness computation (console world) and
circuit synthesis (circuit world), which is necessary to achieve some
optimizations in the circuits.

There is a TODO for the witness computation in `Group::new`: there must be
already code to check whether a pair of coordinates is in the (sub)group. This
TODO is just at the console level; the circuit enforcement is already done in
this commit.

acoglio committed:
Eliminate repeated `//`.

howardwu committed:
Resolve merge

howardwu committed:
Slight comment changes

acoglio committed:
Refine the new refactored group circuits.

In the circuit to enforce that a given point is in the subgroup, tie the witness
values of the internal variables to the witness values of the point. This also
takes care of the TODO that I had initially left.

Adjust variable and constraint counts in tests.

acoglio committed:
fmt

d0cd committed:
Update counts

d0cd committed:
Implement Boolean::assert_bits_are_zero

d0cd committed:
Update counts

d0cd committed:
Apply optimization to casting

d0cd committed:
Merge branch 'tracking/circuit-optimizations-and-fixes' into fix/non-det-in-sqrt

howardwu committed:
Merge pull request #2100 from AleoHQ/optimize/bits-are-zero

[Optimize] Assert bits are zero.

howardwu committed:
Merge pull request #2093 from AleoHQ/optimize/group-circuits

Optimize some group circuits.

howardwu committed:
Merge branch 'tracking/circuit-optimizations-and-fixes' into fix/non-det-in-sqrt

howardwu committed:
Merge pull request #2049 from AleoHQ/fix/non-det-in-sqrt

[Fix] Square root

d0cd committed:
Update counts

d0cd committed:
More updates

raychu86 committed:
Resample parameters and genesis block

raychu86 committed:
Regenerate expectations
